feat(experiments): support multiple primary metrics

jurajmajerik opened the pull request:
## Changes
Support adding/editing/deleting multiple primary metrics + new visualizations.

This is still behind the feature flag and needs more work before rolling out.

**Initial state**
<img width="1383" alt="image" src="https://github.com/user-attachments/assets/e8e3ad7e-b878-41d8-9c4d-93f66c64fcb0" />

**Primary metrics populated + "no results" diagnostics"**
<img width="1377" alt="image" src="https://github.com/user-attachments/assets/cf82ed10-fdc6-4e16-92f4-43c2d5182327" />

**Unexpected error**
<img width="1378" alt="image" src="https://github.com/user-attachments/assets/7b174ab9-0356-4b26-9d03-ed32ef06c052" />

**Dark mode**
<img width="1382" alt="image" src="https://github.com/user-attachments/assets/ea5ec924-3807-44f4-a5f7-a753e0a03baa" />

## How did you test this code?
👀

jurajmajerik committed:
init

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into experiments-primary-metrics

jurajmajerik committed:
wip

jurajmajerik committed:
wip

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into experiments-primary-metrics

jurajmajerik committed:
wip

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into experiments-primary-metrics

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

danielbachhuber reviewed (commented):
Looks like a good start!

No blocking comments on my part, other than it seems like we might want to put this behind a feature flag.

jurajmajerik committed:
wip

jurajmajerik committed:
Merge branch 'experiments-primary-metrics' of https://github.com/PostHog/posthog into experiments-primary-metrics

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into experiments-primary-metrics

jurajmajerik commented on frontend/src/scenes/experiments/ExperimentView/ExperimentView.tsx:
```diff
@@ -49,6 +50,7 @@ const ResultsTab = (): JSX.Element => {
                     )}
                 </>
             )}
+            <MetricsView />
```
Ah, forgot to push my last commit 🙈

jurajmajerik commented on frontend/src/scenes/experiments/MetricsView/DeltaChart.tsx:
```diff
@@ -0,0 +1,653 @@
+                                    ) : (
+                                        <LemonTag size="small" type="danger" className="ml-1">
+                                            Error
+                                        </LemonTag>
```
Interesting, what does `Error` show on hover?

jurajmajerik replied:
> I think I'd expect to see this label appear to the left of the text
Done!

danielbachhuber replied:
> Interesting, what does `Error` show on hover?

I don't see the `Error` right now but I'm seeing an empty tooltip for "Waiting for experiment to start..."

![CleanShot 2024-12-18 at 07 10 25@2x](https://github.com/user-attachments/assets/4ffa300f-db83-4e9e-9335-484a23cb2f99)

jurajmajerik commented on frontend/src/scenes/experiments/MetricsView/DeltaChart.tsx:
```diff
@@ -0,0 +1,653 @@
+                                    const cRate = conversionRateForVariant(result, 'control')
+                                    return vRate && cRate ? (vRate - cRate) / cRate : 0
+                                })
+                            )
```
I've reworked this so the above section has been completely removed.

I've thought more about how the bars should be colored and also checked how our competitors handle this. It turns out that coloring the best bar green doesn’t always make sense, because even the best variant might be entirely in the negative values!

Instead, similar Bayesian tools color the portions of the bar in negative values red and those in positive values green. Here’s how it looks now:

<img width="1379" alt="image" src="https://github.com/user-attachments/assets/29788018-685e-446e-a7fd-843b6c4fe834" />

danielbachhuber replied:
> Here’s how it looks now:

Cool 😄 Worth updating the PR description accordingly?

jurajmajerik replied:
Done

jurajmajerik committed:
wip

jurajmajerik commented:
@danielbachhuber I've addressed your comments, please have one more look.

I've also created an issue to track the todo's before rolling this out: https://github.com/PostHog/posthog/issues/27014

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

danielbachhuber commented on frontend/src/scenes/experiments/MetricsView/DeltaChart.tsx:
```diff
@@ -0,0 +1,660 @@
+                        </div>
+                    </div>
+                </div>
+            </div>
```
I feel like these could use a little more padding:

![CleanShot 2024-12-18 at 07 08 31@2x](https://github.com/user-attachments/assets/5fdcbbaf-4fdd-407c-a026-4edb57d7e833)

jurajmajerik replied:
Fixed

danielbachhuber commented on frontend/src/scenes/experiments/MetricsView/DeltaChart.tsx:
```diff
@@ -0,0 +1,660 @@
+                                // eslint-disable-next-line react/forbid-dom-props
+                                style={{ fontSize: '10px', fontWeight: 400 }}
+                            >
+                                <span>Results loading...</span>
```
```suggestion
                                <span>Results loading&hellip;</span>
```

jurajmajerik replied:
> I don't see the Error right now but I'm seeing an empty tooltip for "Waiting for experiment to start..."

Good catch, fixed.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

jurajmajerik committed:
fixes

jurajmajerik committed:
Merge branch 'experiments-primary-metrics' of https://github.com/PostHog/posthog into experiments-primary-metrics

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into experiments-primary-metrics

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)
